pkg/dns/dnscache.go

```diff
@@ -0,0 +1,432 @@
+
+		if err := updateNftSet(newIPs, deletedIPs, setName, dtype); err != nil {
+			return fmt.Errorf("failed to update nft set: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (e *ipEntry) getNewAndDeletedIPs(ips []net.IP) (newIPs, deletedIPs []nftables.SetElement) {
+	currentIps := make(map[string]bool, len(e.ips))
+	for _, ip := range e.ips {
+		currentIps[ip] = false
+	}
+
+	for _, ip := range ips {
+		s := ip.String()
+		if _, ok := currentIps[s]; ok {
+			currentIps[s] = true
+		} else {
+			newIPs = append(newIPs, nftables.SetElement{Key: ip})
+		}
+	}
+
+	for ip, exists := range currentIps {
+		if !exists {
+			deletedIPs = append(deletedIPs, nftables.SetElement{Key: net.ParseIP(ip)})
+		}
+	}
+
+	return
+}
+
+type cacheEntry struct {
+	ipv4 *ipEntry
+	ipv6 *ipEntry
+}
+
+type DNSCache struct {
+	sync.RWMutex
+
+	log           logr.Logger
+	fqdnToEntry   map[string]cacheEntry
+	setNames      map[string]struct{}
+	dnsServerAddr string
+	ipv4Enabled   bool
+	ipv6Enabled   bool
+}
+
+func NewDNSCache(ipv4Enabled, ipv6Enabled bool, log logr.Logger) *DNSCache {
+	return &DNSCache{
+		log:           log,
+		fqdnToEntry:   map[string]cacheEntry{},
+		setNames:      map[string]struct{}{},
+		dnsServerAddr: defaultDNSServerAddr,
+		ipv4Enabled:   ipv4Enabled,
+		ipv6Enabled:   ipv6Enabled,
+	}
+}
+
+// GetSetsForFQDN returns sets for FQDN selector
+func (c *DNSCache) GetSetsForFQDN(fqdn firewallv1.FQDNSelector, update bool) (result []firewallv1.IPSet) {
+	if update {
+		c.restoreSets(fqdn)
+	}
+
+	sets := map[string]firewallv1.IPSet{}
+	if fqdn.MatchName != "" {
+		for _, s := range c.getSetNameForFQDN(fqdn.GetMatchName()) {
+			sets[s.SetName] = s
+		}
+
+	} else if fqdn.MatchPattern != "" {
+		for _, s := range c.getSetNameForRegex(fqdn.GetRegex()) {
+			sets[s.SetName] = s
+		}
+	}
+
+	result = make([]firewallv1.IPSet, 0, len(sets))
+	for _, s := range sets {
+		result = append(result, s)
+	}
+
+	c.log.WithValues("fqdn", fqdn, "sets", result).Info("sets for FQDN")
+	return
+}
+
+func (c *DNSCache) GetSetsForRendering(fqdns []firewallv1.FQDNSelector) (result []firewallv1.IPSet) {
+	for n, e := range c.fqdnToEntry {
+		var matched bool
+		for _, fqdn := range fqdns {
+			if fqdn.MatchName != "" {
+				if fqdn.GetMatchName() == n {
+					matched = true
+					break
+				}
+			} else if fqdn.MatchPattern != "" {
+				if m, _ := regexp.MatchString(fqdn.GetRegex(), n); m {
+					matched = true
+					break
+				}
+			}
+		}
+		if matched {
+			if e.ipv4 != nil {
+				result = append(result, createIPSetFromIPEntry(n, IPv4, e.ipv4))
+			}
+			if e.ipv6 != nil {
+				result = append(result, createIPSetFromIPEntry(n, IPv6, e.ipv6))
+			}
+		}
+	}
+
+	return
+}
+
+func (c *DNSCache) UpdateDNSServerAddr(addr string) {
+	c.dnsServerAddr = addr
+}
+
+// restoreSets add missing sets from FQDNSelector.Sets
+func (c *DNSCache) restoreSets(fqdn firewallv1.FQDNSelector) {
+	for _, s := range fqdn.Sets {
+		// Add cache entries from fqdn.Sets if missing
+		c.Lock()
```

Comment:
We could consider a sync map for the cache entries to get rid off the complex mutex handling.

Reply:
Problem is that we need 2 maps. One that store set names and another that stores FQDNs. Combining them into one map would be difficult.

Reply:
I agree with @GrigoriyMikhalkin